// angular-paginator/esm5/app/angular-paginator/pipes/angular-paginator.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Pipe } from '@angular/core';
import { AngularPaginatorService } from '../services/angular-paginator.service';
/**
 * Pagination Pipe which paginates the array for the provided arguments such as `itemsPerPage` and `currentPage`
 */
var AngularPaginatorPipe = /** @class */ (function () {
    /**
     *
     * @param _angularPaginatorService service for angular pagiantor
     */
    function AngularPaginatorPipe(_angularPaginatorService) {
        this._angularPaginatorService = _angularPaginatorService;
    }
    /**
     * this pipes paginates the array for the provided arguments such as `itemsPerPage` and `currentPage`
     *
     * @param array input array for which the manipulation happens
     * @param args input arguments for the paginator pipe
     */
    /**
     * this pipes paginates the array for the provided arguments such as `itemsPerPage` and `currentPage`
     *
     * @param {?} array input array for which the manipulation happens
     * @param {?=} args input arguments for the paginator pipe
     * @return {?}
     */
    AngularPaginatorPipe.prototype.transform = /**
     * this pipes paginates the array for the provided arguments such as `itemsPerPage` and `currentPage`
     *
     * @param {?} array input array for which the manipulation happens
     * @param {?=} args input arguments for the paginator pipe
     * @return {?}
     */
    function (array, args) {
        /** @type {?} */
        var instance = this.createInstance(array, args);
        // create pagination information
        this._angularPaginatorService.registerInstance(instance);
        /** @type {?} */
        var start = (instance.currentPage - 1) * instance.itemsPerPage;
        /** @type {?} */
        var end = instance.currentPage * instance.itemsPerPage;
        return array.slice(start, end);
    };
    /**
     * create an instance for each pipe
     *
     * @param array input array for which the manipulation happens
     * @param args input arguments for the paginator pipe
     */
    /**
     * create an instance for each pipe
     *
     * @param {?} array input array for which the manipulation happens
     * @param {?} args input arguments for the paginator pipe
     * @return {?}
     */
    AngularPaginatorPipe.prototype.createInstance = /**
     * create an instance for each pipe
     *
     * @param {?} array input array for which the manipulation happens
     * @param {?} args input arguments for the paginator pipe
     * @return {?}
     */
    function (array, args) {
        return {
            id: args['id'] ? args.id : this._angularPaginatorService.id,
            currentPage: args['currentPage'] ? args['currentPage'] : 1,
            itemsPerPage: args['itemsPerPage'] ? args['itemsPerPage'] : 10,
            totalItems: array['length']
        };
    };
    AngularPaginatorPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'angularPaginator'
                },] }
    ];
    /** @nocollapse */
    AngularPaginatorPipe.ctorParameters = function () { return [
        { type: AngularPaginatorService }
    ]; };
    return AngularPaginatorPipe;
}());
export { AngularPaginatorPipe };
if (false) {
    /** @type {?} */
    AngularPaginatorPipe.prototype._angularPaginatorService;
}
