// angular-paginator/esm5/app/angular-paginator/services/angular-paginator.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable, EventEmitter } from '@angular/core';
var AngularPaginatorService = /** @class */ (function () {
    function AngularPaginatorService() {
        /**
         * id is used to identify pagination instances
         */
        this.id = 'ANGULAR_PAGINATOR_DEFAULT';
        /**
         * instances is the collection of pagination componetns
         */
        this.instances = {};
        /**
         * emits an event with instance id whenever the page number changes
         */
        this.change = new EventEmitter();
    }
    /**
     * creates an instance for every pipe differentiated with id
     *
     * @param instance each instance refers to a unique pagination component
     */
    /**
     * creates an instance for every pipe differentiated with id
     *
     * @param {?} instance each instance refers to a unique pagination component
     * @return {?}
     */
    AngularPaginatorService.prototype.registerInstance = /**
     * creates an instance for every pipe differentiated with id
     *
     * @param {?} instance each instance refers to a unique pagination component
     * @return {?}
     */
    function (instance) {
        if (!this.instances[instance.id]) {
            this.instances[instance.id] = instance;
            this.change.emit(instance.id);
        }
        else {
            /** @type {?} */
            var changed = this.updateInstance(instance);
            if (changed) {
                this.change.emit(instance.id);
            }
        }
        return;
    };
    /**
     * updates existing pagination instances available
     * and returns true if value has changed
     *
     * @param instance an unique pagination component
     */
    /**
     * updates existing pagination instances available
     * and returns true if value has changed
     *
     * @param {?} instance an unique pagination component
     * @return {?}
     */
    AngularPaginatorService.prototype.updateInstance = /**
     * updates existing pagination instances available
     * and returns true if value has changed
     *
     * @param {?} instance an unique pagination component
     * @return {?}
     */
    function (instance) {
        for (var key in instance) {
            if (instance[key] !== this.instances[instance['id']][key]) {
                this.instances[instance['id']][key] = instance[key];
                return true;
            }
        }
        return false;
    };
    /**
     * return instance with given id
     *
     * @param id unique id of the pagination component
     */
    /**
     * return instance with given id
     *
     * @param {?} id unique id of the pagination component
     * @return {?}
     */
    AngularPaginatorService.prototype.getInstance = /**
     * return instance with given id
     *
     * @param {?} id unique id of the pagination component
     * @return {?}
     */
    function (id) {
        return this.instances[id];
    };
    /**
     * return currentPage for given id
     *
     * @param id unique id of the pagination component
     */
    /**
     * return currentPage for given id
     *
     * @param {?} id unique id of the pagination component
     * @return {?}
     */
    AngularPaginatorService.prototype.getCurrentPage = /**
     * return currentPage for given id
     *
     * @param {?} id unique id of the pagination component
     * @return {?}
     */
    function (id) {
        return this.instances[id].currentPage;
    };
    AngularPaginatorService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AngularPaginatorService.ctorParameters = function () { return []; };
    return AngularPaginatorService;
}());
export { AngularPaginatorService };
if (false) {
    /**
     * id is used to identify pagination instances
     * @type {?}
     */
    AngularPaginatorService.prototype.id;
    /**
     * instances is the collection of pagination componetns
     * @type {?}
     */
    AngularPaginatorService.prototype.instances;
    /**
     * emits an event with instance id whenever the page number changes
     * @type {?}
     */
    AngularPaginatorService.prototype.change;
}
